### PR TITLE
feat: create abstract layer for deep_link_client to simplify other implementations

### PR DESCRIPTION
**Pull request by elianortega**
## Description

With the Firebase dynamic link deprecation, the template `DeepLinkClient` should expose an abstract class that enables easy implementation of the deep link service with another package.

## Type of Change

- [ ] ✨ New feature (non-breaking change which adds functionality)
- [ ] 🛠️ Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ Breaking change (fix or feature that would cause existing functionality to change)
- [X] 🧹 Code refactor
- [ ] ✅ Build configuration change
- [ ] 📝 Documentation
- [ ] 🗑️ Chore

### CODE REVIEW

**Comment by elianortega** [comment]
If there is any feedback regarding the naming of the `DeepLinkService` and the updated folder structure for `deep_link_client/`, please let me know 🙌 

I also ran all the workflows on this [PR](https://github.com/elianortega/news_toolkit/pull/2) from my repo.


Finally, what are your thought on providing another implementation of the `DeepLinkService`? Dynamic links are on their way to be deprecated so I think that providing another implementation using [`package:app_links`](https://pub.dev/packages/app_links) or [`package:uni_links`](https://pub.dev/packages/uni_links) could be a greate addition to the template.

**Review by matiasleyba — CHANGES_REQUESTED** [review]

**Comment by matiasleyba** [comment]
> If there is any feedback regarding the naming of the `DeepLinkService` and the updated folder structure for `deep_link_client/`, please let me know 🙌
> 
> I also ran all the workflows on this [PR](https://github.com/elianortega/news_toolkit/pull/2) from my repo.
> 
> Finally, what are your thought on providing another implementation of the `DeepLinkService`? Dynamic links are on their way to be deprecated so I think that providing another implementation using [`package:app_links`](https://pub.dev/packages/app_links) or [`package:uni_links`](https://pub.dev/packages/uni_links) could be a greate addition to the template.



> If there is any feedback regarding the naming of the `DeepLinkService` and the updated folder structure for `deep_link_client/`, please let me know 🙌
> 
> I also ran all the workflows on this [PR](https://github.com/elianortega/news_toolkit/pull/2) from my repo.
> 
> Finally, what are your thought on providing another implementation of the `DeepLinkService`? Dynamic links are on their way to be deprecated so I think that providing another implementation using [`package:app_links`](https://pub.dev/packages/app_links) or [`package:uni_links`](https://pub.dev/packages/uni_links) could be a greate addition to the template.


I agree that adding a new implementation of deeplinking would be great 👍🏻

`app_links` seems to be one of the best packages out there and meets the toolkit requirements. 
`uni_links` seems abandoned, I wouldn't use it.

Remember to keep an eye on the package configuration, as we should configure both Android and iOS platforms in the template and also document any necessary configuration needed.

File: flutter_news_example/packages/deep_link_client/deep_link_client/lib/src/deep_link_service.dart [hunk 1]
```diff
@@ -0,0 +1,11 @@
+/// {@template deep_link_service}
+/// A generic deep link service interface.
+/// {@endtemplate}
+abstract class DeepLinkService {
```
[the remark this answers is not in the document]
Comment (elianortega) [reply flutter_news_example/packages/deep_link_client/deep_link_client/lib/src/deep_link_service.dart:4]: As we already have the `DeepLinkClient`, do you mean renaming it to `DeepLinkServiceClient` or what did you have in mind?

File: flutter_news_example/packages/deep_link_client/firebase_deep_link_service/lib/src/firebase_deep_link_service.dart [hunk 2]
```diff
@@ -0,0 +1,26 @@
+import 'dart:async';
+
+import 'package:deep_link_client/deep_link_client.dart';
+import 'package:firebase_dynamic_links/firebase_dynamic_links.dart';
+
+/// {@template firebase_deep_link_client}
+/// A FirebaseDynamicLinks implementation of [DeepLinkService].
+/// {@endtemplate}
+class FirebaseDeepLinkService extends DeepLinkService {
```
[the remark this answers is not in the document]
Comment (elianortega) [reply flutter_news_example/packages/deep_link_client/firebase_deep_link_service/lib/src/firebase_deep_link_service.dart:9]: I agree, it should be implement.

File: flutter_news_example/packages/deep_link_client/firebase_deep_link_service/lib/src/firebase_deep_link_service.dart [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+  /// {@macro firebase_deep_link_client}
+  FirebaseDeepLinkService({FirebaseDynamicLinks? firebaseDynamicLinks})
+      : _firebaseDynamicLinks =
+            firebaseDynamicLinks ?? FirebaseDynamicLinks.instance;
+
+  late final FirebaseDynamicLinks _firebaseDynamicLinks;
```
[the remark this answers is not in the document]
Comment (elianortega) [reply flutter_news_example/packages/deep_link_client/firebase_deep_link_service/lib/src/firebase_deep_link_service.dart:15]: Yup, not needed

**Comment by elianortega** [comment]
> @elianortega I agree that adding a new implementation of deeplinking would be great 👍🏻
> 
> `app_links` seems to be one of the best packages out there and meets the toolkit requirements. `uni_links` seems abandoned, I wouldn't use it.
> 
> Remember to keep an eye on the package configuration, as we should configure both Android and iOS platforms in the template and also document any necessary configuration needed.

Yes, I'll send that to another PR after this one gets merged 👍

File: flutter_news_example/packages/deep_link_client/deep_link_client/lib/src/deep_link_service.dart (hunk 1 above)
[the remark this answers is not in the document]
Comment (matiasleyba) [reply flutter_news_example/packages/deep_link_client/deep_link_client/lib/src/deep_link_service.dart:4]: I mean change `DeepLinkService` to `DeepLinkClient`.

[the remark this answers is not in the document]
Comment (matiasleyba) [reply flutter_news_example/packages/deep_link_client/deep_link_client/lib/src/deep_link_service.dart:4]: We can exchange names between the service and the client, something like this: 

```
DeepLinkService({
  required DeepLinkClient deepLinkClient,
})
```

What do you think?

[the remark this answers is not in the document]
Comment (elianortega) [reply flutter_news_example/packages/deep_link_client/deep_link_client/lib/src/deep_link_service.dart:4]: I agree, I'll make the update 👍

File: flutter_news_example/packages/deep_link_client/deep_link_client/lib/src/deep_link_service.dart [hunk 4]
```diff
@@ -0,0 +1,11 @@
+/// {@template deep_link_client}
```
Comment (matiasleyba) [inline flutter_news_example/packages/deep_link_client/deep_link_client/lib/src/deep_link_service.dart:1]: File name should be updated

File: flutter_news_example/packages/deep_link_client/deep_link_client/lib/src/deep_link_client.dart [hunk 5]
```diff
@@ -1,8 +1,7 @@
 import 'dart:async';
```
Comment (matiasleyba) [inline flutter_news_example/packages/deep_link_client/deep_link_client/lib/src/deep_link_client.dart:1]: same here

File: flutter_news_example/packages/deep_link_client/firebase_deep_link_client/lib/src/firebase_deep_link_client.dart [hunk 6]
```diff
@@ -0,0 +1,26 @@
+import 'dart:async';
+
+import 'package:deep_link_client/deep_link_client.dart';
+import 'package:firebase_dynamic_links/firebase_dynamic_links.dart';
+
+/// {@template firebase_deep_link_client}
+/// A FirebaseDynamicLinks implementation of [DeepLinkService].
```
Comment (matiasleyba) [inline flutter_news_example/packages/deep_link_client/firebase_deep_link_client/lib/src/firebase_deep_link_client.dart:7]: typo: `A FirebaseDynamicLinks implementation of [DeepLinkService].` > `A FirebaseDynamicLinks implementation of [DeepLinkClient].`

**Review by matiasleyba — CHANGES_REQUESTED** [review]

**Review by matiasleyba — CHANGES_REQUESTED** [review]
@elianortega Thanks for addressing the comments, the changes look good! Anyway I would like to hold the merge until we solve a issue we are having with the CI workflows.
Once we solve the issue we will be able to merge the PR, I will keep you posted on the issue.

**Comment by elianortega** [comment]
No worries @matiasleyba, let me know if I can help with anything 👍

**Comment by matiasleyba** [comment]
Hi @elianortega, we have been able to resolve the issues with the CI workflows, would you mind updating the PR? Thanks in advance!

**Review by matiasleyba — CHANGES_REQUESTED** [review]
Hi @elianortega , reviewing the project again I see that the `deep_link_client` is very much tied to the email link authentication functionality and not to a deep linking solution as the current navigation approach does not support deep links.
Basically we are using dynamic links because of the email link authentication functionality, but this functionality will not be deprecated, it will be [replaced](https://firebase.google.com/support/dynamic-links-faq#i_need_to_onboard_onto_firebase_dynamic_links_to_enable_email_link_auth_in_firebase_authentication_what_should_i_do).

Do you think it is still worth making the change, what do you think?

**Comment by elianortega** [comment]
Hey @matiasleyba sorry for the late reply, 

>reviewing the project again I see that the deep_link_client is very much tied to the email link authentication functionality and not to a deep linking solution

Yes, the current scope in the toolkit and for this change was just for authentication purposes, but I think adding the `DeepLinkClient` abstraction is can become a feature for the toolkit. From my experience using the toolkit twice, I ended up adding deep links anyway, as it is a very requested feature for news applications. 

But maybe this PR could be the beginning of adding deep link support to the toolkit. To handle this appropriately I'll create a feature request for the toolkit, and resolve the conflicts of this PR.

**Comment by matiasleyba** [comment]
> Hey @matiasleyba sorry for the late reply,
> 
> > reviewing the project again I see that the deep_link_client is very much tied to the email link authentication functionality and not to a deep linking solution
> 
> Yes, the current scope in the toolkit and for this change was just for authentication purposes, but I think adding the `DeepLinkClient` abstraction is can become a feature for the toolkit. From my experience using the toolkit twice, I ended up adding deep links anyway, as it is a very requested feature for news applications.
> 
> But maybe this PR could be the beginning of adding deep link support to the toolkit. To handle this appropriately I'll create a feature request for the toolkit, and resolve the conflicts of this PR.

It would be great to add support for deep linking, but we should address this by changing the navigation approach first, probably using go_router.

**Comment by matiasleyba** [comment]
@elianortega As the implementation of the DeepLinkClient class is very much tied to Firebase auth I don't see another use case for this class apart from authentication using a link, I would like to know if you find any other use case or a reason to abstract it ?

**Comment by elianortega** [comment]
If there are no issues with moving towards another navigation approach, I agree that refactoring to use go_router first is the way to go, that would also enable the usage of the built-in deep link handling from go_router for basic navigation. Is this change already considered in the backlog of the toolkit?

> I would like to know if you find any other use case or a reason to abstract it?

The only use case I found for the abstraction, is to be able to quickly implement the deep link handling with another package like: app_links or flutter_branch_sdk. In some experiences I've had with the toolkit, even we change to go router, we still used a custom deep link implementation to do extra work with the deep link.

But If the migration to go_router is already in the plan, I think that would cover the must common scenario of navigation to news article.

**Comment by matiasleyba** [comment]
> If there are no issues with moving towards another navigation approach, I agree that refactoring to use go_router first is the way to go, that would also enable the usage of the built-in deep link handling from go_router for basic navigation. Is this change already considered in the backlog of the toolkit?
> 
> > I would like to know if you find any other use case or a reason to abstract it?
> 
> The only use case I found for the abstraction, is to be able to quickly implement the deep link handling with another package like: app_links or flutter_branch_sdk. In some experiences I've had with the toolkit, even we change to go router, we still used a custom deep link implementation to do extra work with the deep link.
> 
> But If the migration to go_router is already in the plan, I think that would cover the must common scenario of navigation to news article.

Deeplinking is in the plan, but we don't have an estimate of when we will start working on it.

As for abstraction, if you find it useful in previous implementations, we can continue working on this PR.
I will keep an eye on the PR when it is ready for review.

File: flutter_news_example/packages/deep_link_client/firebase_deep_link_client/test/firebase_deep_link_client_test.dart [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+            FirebaseDeepLinkClient(firebaseDynamicLinks: dynamicLinks);
+
+        expect(
+          client.deepLinkStream,
+          emitsInOrder(
+            <Uri>[expectedUri1, expectedUri1, expectedUri2, expectedUri1],
+          ),
+        );
+
+        onLinkStreamController
```
Comment (matiasleyba) [inline flutter_news_example/packages/deep_link_client/firebase_deep_link_client/test/firebase_deep_link_client_test.dart:63]: I think this should be above `expect`, what do you think ?

File: flutter_news_example/packages/deep_link_client/firebase_deep_link_client/test/firebase_deep_link_client_test.dart [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+          ..add(PendingDynamicLinkData(link: expectedUri1))
+          ..add(PendingDynamicLinkData(link: expectedUri1))
+          ..add(PendingDynamicLinkData(link: expectedUri2))
+          ..add(PendingDynamicLinkData(link: expectedUri1));
+      });
+    });
+
+    group('with default FirebaseDynamicLinks', () {
```
Comment (matiasleyba) [inline flutter_news_example/packages/deep_link_client/firebase_deep_link_client/test/firebase_deep_link_client_test.dart:71]: by making `firebaseDynamicLinks` required we can avoid writing those tests, you can check a similar implementation at `FirebaseNotificationsClient`

**Review by matiasleyba — CHANGES_REQUESTED** [review]
@elianortega Just a few comments before merge

File: flutter_news_example/packages/deep_link_client/firebase_deep_link_client/pubspec.yaml [hunk 9]
```diff
@@ -1,16 +1,18 @@
...
 
 dependencies:
+  deep_link_client:
+    path: ../deep_link_client
   equatable: ^2.0.3
   firebase_core: ^3.0.0
   firebase_dynamic_links: ^6.0.0
   plugin_platform_interface: ^2.1.3
-  rxdart: ^0.27.3
+  rxdart: ^0.27.5
```
[the remark this answers is not in the document]
Comment (elianortega) [reply flutter_news_example/packages/deep_link_client/firebase_deep_link_client/pubspec.yaml:15]: With the latest main it is so it matches all the other rxdart versions in the project.

File: flutter_news_example/packages/deep_link_client/firebase_deep_link_client/test/firebase_deep_link_client_test.dart (hunk 7 above)
Comment (elianortega) [reply flutter_news_example/packages/deep_link_client/firebase_deep_link_client/test/firebase_deep_link_client_test.dart:63]: I agree

File: flutter_news_example/packages/deep_link_client/firebase_deep_link_client/test/firebase_deep_link_client_test.dart (hunk 8 above)
Comment (elianortega) [reply flutter_news_example/packages/deep_link_client/firebase_deep_link_client/test/firebase_deep_link_client_test.dart]: I've made the change to match the standard

**Comment by elianortega** [comment]
@matiasleyba Thanks for the review, I've made the changes, please let me know if you have any other feedback on this one

File: flutter_news_example/packages/deep_link_client/firebase_deep_link_client/pubspec.yaml (hunk 9 above)
[the remark this answers is not in the document]
Comment (matiasleyba) [reply flutter_news_example/packages/deep_link_client/firebase_deep_link_client/pubspec.yaml:15]: That makes sense, but it seems that dependency is not necessary for this package, am I right?

**Review by matiasleyba — CHANGES_REQUESTED** [review]
@elianortega thanks for addressing changes, just one pending before merging.

**Comment by elianortega** [comment]
@matiasleyba It is true, it was not being used my bad, I removed it

**Review by matiasleyba — APPROVED** [review]

**Pull request merged**